src/test/java/test/thread/parallelization/BaseParallelizationTest.java

```diff
@@ -1,30 +1,591 @@
+                            "onStart event");
+                    assertTrue(suitesExecuting.get(suiteName).getThreadId() == eventLog.getThreadId(), "All the " +
+                            "suite level event logs for a given suite should have the same thread ID");
+
+                    suitesExecuting.remove(suiteName);
+                    executingSuiteThreadIds.remove(eventLog.getThreadId());
+                    suitesCompleted.put((String)eventLog.getData(SUITE_NAME),eventLog);
+                }
+            }
+        }
+    }
+
+    public static void verifySimultaneousSuiteListenerStartEvents(List<EventLog> listenerStartEventLogs, int
+            threadPoolSize) {
+
+        verifyEventTypeForEventsLogs(listenerStartEventLogs, LISTENER_SUITE_START, "The suite thread pool size is " +
+                threadPoolSize + ", so no more more than " + threadPoolSize + " suites should start running at the " +
+                "the same time if there are more than " + threadPoolSize + " suites remaining to execute.");
+        verifyDifferentThreadIdsForEvents(listenerStartEventLogs, "The suite thread pool size is " + threadPoolSize +
+                ", so the thread IDs for all the suite listener's onStart method for the " + threadPoolSize +
+                " currently executing suites should be different");
+    }
+
+    private static void verifyFirstBlockOfSimultaneouslyExecutingSuites(List<EventLog> suiteListenerStartEventLogs,
+            int threadPoolSize) {
+
+        verifySimultaneousSuiteListenerStartEvents(suiteListenerStartEventLogs, threadPoolSize);
+        verifyDifferentThreadIdsForEvents(suiteListenerStartEventLogs, "The thread count is " + threadPoolSize +
+                " so the thread IDs for the suite listener onStart events for simultaneously executing suites " +
+                "should be different. Event logs: " + suiteListenerStartEventLogs);
+    }
+
+    //Helper which verifies that the specified lists of test method level events are associated with the same methods
+    private static void verifyEventsBelongToSameMethods(List<EventLog> firstEventLogs, List<EventLog> secondEventLogs,
+            String faiMessage) {
+
+        List<String> methodNames = new ArrayList<>();
+
+        for(EventLog eventLog : firstEventLogs) {
+            methodNames.add((String)eventLog.getData(CLASS_NAME) + (String)eventLog.getData(METHOD_NAME));
+        }
+
+        for(EventLog eventLog : secondEventLogs) {
+
+            assertTrue(methodNames.contains((String)eventLog.getData(CLASS_NAME) +
+                    (String)eventLog.getData(METHOD_NAME)), faiMessage);
+        }
+    }
+
+    //Helper method that retrieves the earliest and latest timestamps for the specified list of event logs
+    private static Pair<Long, Long> getEarliestAndLatestTimestamps(List<EventLog> eventLogs) {
+
+        if(!eventLogs.isEmpty()) {
```

Comment:
Personal opinion, but most of the time I prefer quick exit because it highlights the important code branch:

```java
if(eventLogs.isEmpty()) {
  return null;
}
```